explorer/interpreter/type_checker.cpp

```diff
@@ -2863,6 +2863,24 @@ auto TypeChecker::TypeCheckStmt(Nonnull<Statement*> s,
       CARBON_RETURN_IF_ERROR(TypeCheckStmt(&while_stmt.body(), impl_scope));
       return Success();
     }
+    case StatementKind::For: {
+      auto& for_stmt = cast<For>(*s);
+      ImplScope inner_impl_scope;
+      inner_impl_scope.AddParent(&impl_scope);
```

Comment:
I don't think we should need a separate `ImplScope` here, since `impl`s can't be nested within the `for` statement.

Reply:
Please corrent me, but isn't it so:

```
Block
for( x: i32 in ar) //new Block{
//new Block
}
Block
```
Because after the execution of the loop, *x* shouldn't be found in Block.

Reply:
Another problem is  that TypeCheckPattern expected a non const ImplScope,  but impl_scope is const!